### PR TITLE
Override submitZKPResponxeV2 for Embeddedzkpverifier

### PR DESCRIPTION
**Pull request by daveroga**
Override the submitZKPResponseV2 for EmbeddedZKPVerifier

- Add the _beforeProofSubmitV2 and _afterProofSubmitV2 hooks.
- Need to figure out the interfaces for them. 
- Add the unit tests to `embedded-zkp-verifier.test.ts`.

### CODE REVIEW

File: test/verifier/embedded-zkp-verifier.test.ts [hunk 1]
```diff
@@ -2,11 +2,18 @@ import { expect } from "chai";
...
-import { prepareInputs, publishState } from "../utils/state-utils";
+import { prepareInputs } from "../utils/state-utils";
 import { Block, Signer } from "ethers";
+import {
+  packCrossChainProofs,
+  packGlobalStateUpdateWithSignature,
+  packIdentityStateUpdateWithSignature,
+  packZKProof,
+} from "../utils/packData";
+import proofJson from "../validators/sig/data/valid_sig_user_genesis.json";
```
Comment (AndriianChestnykh) [inline test/verifier/embedded-zkp-verifier.test.ts:13]: As long as you have Groth16VerifierValidatorStub now you can put any public inputs without invalidation groth16 proof. 
In That way you can avoid messages like "id type is not supported",  "Challenge should match the sender", ...

File: test/verifier/embedded-zkp-verifier.test.ts [hunk 2]
```diff
@@ -70,6 +75,83 @@ describe("Embedded ZKP Verifier", function () {
...
+      },
+      {
+        proofType: "stateProof",
+        proof: await packIdentityStateUpdateWithSignature(identityStateMessage1, signer),
+      },
+      {
+        proofType: "stateProof",
+        proof: await packIdentityStateUpdateWithSignature(identityStateUpdate2, signer),
+      },
+    ]);
```
Comment (AndriianChestnykh) [inline test/verifier/embedded-zkp-verifier.test.ts:123]: If not mistaken, this code repeats in many placed of the repo. Consider replacing with some helper function

Comment (daveroga) [reply test/verifier/embedded-zkp-verifier.test.ts]: I created a helper function to build the proofs.

File: test/utils/packData.ts [hunk 3]
```diff
@@ -145,3 +145,24 @@ export function packZKProof(inputs: string[], a: string[], b: string[][], c: str
...
+        return {
+          proofType: "globalStateProof",
+          proof: await packGlobalStateUpdateWithSignature(crossChainProofMessage, signer),
+        };
+      }
+      return {
+        proofType: "stateProof",
+        proof: await packIdentityStateUpdateWithSignature(crossChainProofMessage, signer),
+      };
+    }),
```
Comment (AndriianChestnykh) [inline test/utils/packData.ts:165]: simplify with ternary?

File: test/verifier/embedded-zkp-verifier.test.ts [hunk 4]
```diff
@@ -52,8 +53,146 @@ describe("Embedded ZKP Verifier", function () {
...
     const { inputs, pi_a, pi_b, pi_c } = prepareInputs(proofJson);
+
     const tx = await verifier.submitZKPResponse(0, inputs, pi_a, pi_b, pi_c);
     const txRes = await tx.wait();
+    const receipt = await ethers.provider.getTransactionReceipt(txRes.hash);
+
+    // 2 events are emitted
+    expect(receipt?.logs.length).to.equal(2);
+
+    const dataEvent1 = receipt?.logs[0].data;
```
Comment (AndriianChestnykh) [inline test/verifier/embedded-zkp-verifier.test.ts:64]: you can call the vars directly as dataBeforeProof / dataAfterProof ... as bit more readability IMHO

**Review by AndriianChestnykh — APPROVED** [review]

**Pull request merged**